### PR TITLE
Implement misc missing protocol objects into packet events

### PR DESCRIPTION
**Pull request by Tofaa2**
Let me know if it should be structured differently or what to change. I believe these should be in the API rather than magic values (for example VillagerLevel, and Color and DyeColor). Along side brings back Rotation

### CODE REVIEW

**Comment by Tofaa2** [comment]
@retrooper

**Comment by retrooper** [comment]
I like the VillagerLevel and how you didn't want to break compatibility, but the getLevelEnum() brings a frown to my face.

**Comment by retrooper** [comment]
There probably isn't a better solution.

**Comment by Tofaa2** [comment]
> I like the VillagerLevel and how you didn't want to break compatibility, but the getLevelEnum() brings a frown to my face.

Yeah unfortunately I didnt know of a better variable name ngl

**Comment by retrooper** [comment]
Add appropriate copyright notices and it's good to go. 
https://github.com/retrooper/packetevents/wiki/Pull-Request-Requirements

**Comment by retrooper** [comment]
This might sound pedantic, but we add the license notice at the very top of the file.

File: api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/level/VillagerLevel.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    private static final VillagerLevel[] VALUES = values();
+
+    public static VillagerLevel getById(int id) {
+        return VALUES[id];
+    }
+
+    public int getId() {
+        return ordinal();
+    }
+
```
Comment (NoJokeFNA) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/level/VillagerLevel.java:37]:
```suggestion

```

File: api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/VillagerData.java [hunk 2]
```diff
@@ -55,10 +60,19 @@ public void setProfession(VillagerProfession profession) {
...
     }
 
     public void setLevel(int level) {
+        this.level = VillagerLevel.getById(level);
+    }
+
+    public void setLevel(VillagerLevel level) {
         this.level = level;
     }
+
```
Comment (NoJokeFNA) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/VillagerData.java:77]:
```suggestion

```

File: api/src/main/java/com/github/retrooper/packetevents/protocol/color/DyeColor.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+public enum DyeColor implements RGBLike {
+    WHITE(new Color(0xf9fffe), new Color(0xffffff), new Color(0xf0f0f0), 8),
```
Comment (NoJokeFNA) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/color/DyeColor.java:25]: And why is there always an empty line between each enum type, that doesn't necessarily make it more readable for me, rather more filled and therefore more confusing

**Review by NoJokeFNA — CHANGES_REQUESTED** [review]

**Review by NoJokeFNA — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/github/retrooper/packetevents/protocol/color/DyeColor.java (hunk 3 above)
Comment (Tofaa2) [reply api/src/main/java/com/github/retrooper/packetevents/protocol/color/DyeColor.java:25]: I fixed it just for you pookiebear

Comment (NoJokeFNA) [reply api/src/main/java/com/github/retrooper/packetevents/protocol/color/DyeColor.java:25]: Either my eyes are bad or it's still the same

File: api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/level/VillagerLevel.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+}
+
```
Comment (NoJokeFNA) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/level/VillagerLevel.java:40]:
```suggestion

```

File: api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/level/VillagerLevel.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    private static final VillagerLevel[] VALUES = values();
+
+    public static VillagerLevel getById(int id) {
+        return VALUES[id];
+    }
+
+    public int getId() {
+        return ordinal();
+    }
+
```
Comment (NoJokeFNA) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/entity/villager/level/VillagerLevel.java:38]:
```suggestion

```

**Review by NoJokeFNA — CHANGES_REQUESTED** [review]

**Comment by retrooper** [comment]
Thanks for the PR

**Pull request merged**